### ISSUE TITLE
Replication design discussions

### ISSUE THREAD

**Issue by Brar** [Brar/npgsql#2]
Hey @roji,

I'm currently thinking about extracting the streaming replication logic into a class that's separate from NpgsqlReplicationConnection (an explicit `IAsyncEnumerator<T>` implementation instead of an implicit one).

Instead of looking like this:
```csharp
public class NpgsqlReplicationConnection
{
    string ConnectionString { get; set; }
    NpgsqlLogSequenceNumber LastReceivedLsn { get; }
    NpgsqlLogSequenceNumber LastFlushedLsn { get; set; }
    NpgsqlLogSequenceNumber LastAppliedLsn { get; set; }
    TimeSpan WalReceiverStatusInterval { get; set; }
    TimeSpan WalReceiverTimeout { get; set; }
    Task OpenAsync(CancellationToken cancellationToken = default);
    ValueTask DisposeAsync();
    Task<NpgsqlReplicationSystemIdentification> IdentifySystem(CancellationToken cancellationToken = default);
    Task<string> Show(string parameterName, CancellationToken cancellationToken = default);
    Task<NpgsqlTimelineHistoryFile> TimelineHistory(uint tli, CancellationToken cancellationToken = default);
    Task SendStatusUpdate(CancellationToken cancellationToken = default);
    Task<NpgsqlReplicationSlot> CreateReplicationSlot(string slotName,bool temporary = false);
    Task DropReplicationSlot(string slotName, bool wait = false, CancellationToken cancellationToken = default);
    Task<IAsyncEnumerator<NpgsqlXLogDataMessage>> StartReplication(NpgsqlReplicationSlot slot, NpgsqlLogSequenceNumber walLocation, uint timeline = default, CancellationToken cancellationToken = default);
    void Cancel();
}
```

it would look like this:
```csharp
public interface INpgsqlReplicationConnection
{
    string ConnectionString { get; set; }
    Task OpenAsync(CancellationToken cancellationToken = default);
    ValueTask DisposeAsync();
    Task<NpgsqlReplicationSystemIdentification> IdentifySystem(CancellationToken cancellationToken = default);
    Task<string> Show(string parameterName, CancellationToken cancellationToken = default);
    Task<NpgsqlTimelineHistoryFile> TimelineHistory(uint tli, CancellationToken cancellationToken = default);
    Task<NpgsqlReplicationSlot> CreateReplicationSlot(string slotName, bool temporary = false);
    Task DropReplicationSlot(string slotName, bool wait = false, CancellationToken cancellationToken = default);
    Task<NpgsqlReplicationStream> StartReplication(NpgsqlReplicationSlot slot, NpgsqlLogSequenceNumber walLocation, uint timeline = default, CancellationToken cancellationToken = default);
}
public class NpgsqlReplicationStream : IAsyncEnumerator<NpgsqlXLogDataMessage>
{
    NpgsqlLogSequenceNumber LastReceivedLsn { get; }
    NpgsqlLogSequenceNumber LastFlushedLsn { get; set; }
    NpgsqlLogSequenceNumber LastAppliedLsn { get; set; }
    TimeSpan WalReceiverStatusInterval { get; set; }
    TimeSpan WalReceiverTimeout { get; set; }
    ValueTask DisposeAsync();
    Task SendStatusUpdate(CancellationToken cancellationToken = default);
    void Cancel();
}
```

What do you think?

**Comment by Brar** [comment]
Being a setting that you probably never change at runtime I'd probably also move `WalReceiverStatusInterval` and `WalReceiverTimeout` to the connection string where it obviously would only be relevant for replication connections.

**Comment by roji** [comment]
tl;dr I'd not expose a public enumerator type to users - I'm not sure there's an actual advantage, and it would likely make the API harder to work with. Specific reasoning:

* The enumerator API is generally (purposefully) hidden from API consumers, who typically don't manipulate it directly. Forcing users to deal with enumerators directly makes for a difficult API (e.g. they can't use foreach easily), so it's really best avoided (in fact, see last note below about the lack of an IAsyncEnumerable). Enumerators are typically exposed to users only in cases where mutation is supported (e.g. remove the current element), or similar.
* For the actual APIs and properties, since replication is an exclusive connection activity (nothing else can happen), I think they can live just as well on the connection as on the enumerator. If we needed to support multiple active replications on the same connection, then putting them on the enumerator (or on something else outside the connection) would indeed have been necessary. But as things are, I don't see an advantage in doing this (btw, the Last*Lsn properties should probably throw InvalidOperationException if a replication isn't in progress?).
* For the settings specifically (status interval, timeout), I think these fit better on the connection, since I'm assuming users would set them once at startup, and then expect them to persist across multiple replication sessions (if anyone ever does that :roll_eyes:). If they're on the IAsyncEnumerable, you need to set them again each time you start a replication session.

On a related note... I'm not sure from our previous discussions anymore, but shouldn't we be returning an IAsyncEnumerable from StartReplication? IAsyncEnumerable is required to do foreach, and also to compose LINQ operators (e.g. Where)...

**Comment by Brar** [comment]
> shouldn't we be returning an IAsyncEnumerable from StartReplication?

Yeah, don't focus on the `IAsyncEnumerator<T>` vs. `IAsyncEnumerable<T>` question too much, the differenc is minor IMHO.
My reasoning was that the `IAsyncEnumerable<T>` adds another level of indirection and that it isn't meant to be disposable, whereas the `IAsyncEnumerator<T>` is.
But after all that's just an implementation detail and the main question is whether we want an explicit implementation that holds additional properties/methods.

I started thinking about this when I recognized that the `Last*Lsn` properties, the `SendStatusUpdate()` method and the `Cancel()` method only have a meaning while the connection is in streaming mode and would thwor exceptions in all other states.

The Idea was to only make them accessible while the connection is actually streaming.
We might even get rid of `Cancel()` in that case and use `DisposeAsync()` on the returned Stream for cancelling the streaming.

See below for 3 Variants of the API and how it could be used IRL.
This probably makes clearer how I came to this.
As it currently stands (Variant 3) you always have to carry around the stream, the connection and possibly a cancellation token:

```csharp
public async Task Main()
{
    using var cts = new CancellationTokenSource();
    var replicationConnection = new NpgsqlPhysicalReplicationConnection();
    await replicationConnection.OpenAsync(cts.Token);
    var info = await replicationConnection.IdentifySystem(cts.Token);
    var replicationStream = await replicationConnection.StartReplication(info.XLogPos, cancellationToken: cts.Token);

    // Variant 1:
    // replicationStream is a custom class implementing IAsyncEnumerator<NpgsqlXLogDataMessage>
    // and exposing the feedback API
    var replicationThreadTask = Task.Run(() => StreamingLoop(replicationStream));
    // Variant 2:
    // replicationStream is a custom class implementing IAsyncEnumerable<NpgsqlXLogDataMessage>
    // and exposing the feedback API
    //var replicationThreadTask = Task.Run(() => StreamingLoop(replicationStream, cts.Token));
    // Variant 3:
    // replicationStream is an automatic IAsyncEnumerable<NpgsqlXLogDataMessage>,
    // the feedback API is exposed by the connection
    //var replicationThreadTask = Task.Run(() => StreamingLoop(replicationConnection, replicationStream, cts.Token));

    Console.Write("Please enter 'q' to stop replication: ");
    while (true)
    {
        var key = Console.ReadKey(true);
        if (key.KeyChar != 'q')
            continue;

        // Variant 1: IAsyncEnumerator<T> implements IAsyncDisposable<T> which we can use to end the replication stream
        await replicationStream.DisposeAsync();

        // Variant 2 a/b/c: IAsyncEnumerable<T> does not implement IAsyncDisposable<T> so we'd either implement it
        //  in addition which is not recommendable because it would interfere with await foreach or we'd dispose via the
        //  returnedIAsyncEnumerator<T> (which needs to be a single object per connection anyways) or we'd add Cancel()
        //  to the API (leaving the question what happens if our consumer disposes the IAsyncEnumerator - probably a no-op)
        // a:
        //await replicationStream.DisposeAsync();
        // b:
        //await replicationStream.GetAsyncEnumerator().DisposeAsync();
        // c:
        //await replicationStream.Cancel();

        // Variant 3: We probably need Cancel() on the connection
        //await replicationConnection.Cancel();

        // Variant 1, 2 and 3: All three variants would support stopping the replication stream via a cancellation token
        // but the way we need to pass it would be different
        // cts.Cancel();

        try
        {
            await replicationThreadTask;
        }
        catch (OperationCanceledException)
        {
            await Console.Out.WriteLineAsync("Replication was stopped.");
        }
        catch (Exception e)
        {
            await Console.Error.WriteLineAsync($"An exception occurred in the replication thread: {e}");
        }
        break;
    }
    // resume using replicationConnection here as it has ended streaming mode and should be useable again.
}

// Variant 1: replicationStream is a custom class implementing IAsyncEnumerator<NpgsqlXLogDataMessage>
async Task StreamingLoop(NpgsqlReplicationStream replicationStream)
{
    using var feedbackTimeoutCts = new TimeoutCancellationTokenSourceWrapper(TimeSpan.FromSeconds(5));
    while (await replicationStream.MoveNextAsync())
    {
        var msg = replicationStream.Current;
        // Do whatever we want with the NpgsqlXLogDataMessage in msg
        // then update replicationStream.LastAppliedLsn and replicationStream.LastFlushedLsn
        replicationStream.LastAppliedLsn = msg.WalEnd;
        replicationStream.LastFlushedLsn = msg.WalEnd;
        // Optionally enforce immediate feedback using replicationStream.SendStatusUpdate()
        await replicationStream.SendStatusUpdate(feedbackTimeoutCts.Start());
        feedbackTimeoutCts.Stop();
    }
}

// Variant 2: replicationStream is a custom class implementing IAsyncEnumerable<NpgsqlXLogDataMessage>
async Task StreamingLoop(NpgsqlReplicationStream replicationStream, CancellationToken cancellationToken = default)
{
    using var feedbackTimeoutCts = new TimeoutCancellationTokenSourceWrapper(TimeSpan.FromSeconds(5));
    await foreach (var msg in replicationStream.WithCancellation(cancellationToken))
    {
        // Do whatever we want with the NpgsqlXLogDataMessage in msg
        // then update replicationStream.LastAppliedLsn and replicationStream.LastFlushedLsn
        replicationStream.LastAppliedLsn = msg.WalEnd;
        replicationStream.LastFlushedLsn = msg.WalEnd;
        // Optionally enforce immediate feedback using replicationStream.SendStatusUpdate()
        await replicationStream.SendStatusUpdate(feedbackTimeoutCts.Start());
        feedbackTimeoutCts.Stop();
    }
}

// Variant 3: replicationStream is an automatic IAsyncEnumerable<NpgsqlXLogDataMessage>
async Task StreamingLoop(NpgsqlPhysicalReplicationConnection connection,
    IAsyncEnumerable<NpgsqlXLogDataMessage> replicationStream, CancellationToken cancellationToken = default)
{
    using var feedbackTimeoutCts = new TimeoutCancellationTokenSourceWrapper(TimeSpan.FromSeconds(5));
    await foreach (var msg in replicationStream.WithCancellation(cancellationToken))
    {
        // Do whatever we want with the NpgsqlXLogDataMessage in msg
        // then update connection.LastAppliedLsn and connection.LastFlushedLsn
        connection.LastAppliedLsn = msg.WalEnd;
        connection.LastFlushedLsn = msg.WalEnd;
        // Optionally enforce immediate feedback using connection.SendStatusUpdate()
        await connection.SendStatusUpdate(feedbackTimeoutCts.Start());
        feedbackTimeoutCts.Stop();
    }
}
```

**Comment by roji** [comment]
>  We might even get rid of Cancel() in that case and use DisposeAsync() on the returned Stream for cancelling the streaming.

Isn't cancellation handled via the cancellation token given to StartReplication? We don't support any sync I/O here, so I'm not sure we need a Cancel method...

> The Idea was to only make them accessible while the connection is actually streaming.

Yeah, I understand the sentiment. But there's lots of precedent of just having properties which throw InvalidOperationException when not in the right state (e.g. various properties on NpgsqlConnection which report on server details). At the end of the day, I think it's really a question of whether the additional type simplifies the API or makes it more difficult - I think it does the latter in this case.

Re the 3 variants... First, I'd exclude the the IAsyncEnumerator one because it doesn't support await foreach. Now, the way I'd expect 95% of users to consume this API, is as follows:

```c#
foreach (var msg in await conn.StartReplication(info.XLogPos, token))
{
    connection.LastAppliedLsn = msg.WalEnd;
    ...
}
```

In this scenario, neither the enumerable nor the enumerator are accessible to the user, for setting the LSNs etc. I think that on its own is reason enough to put the APIs on the connection rather than on the enumerable.

But I understand what you're seeing - in a scenario where the user wants to pass the enumerable around functions, they also need to pass around the connection. To address this, we could just have the best of both worlds: have StartReplication expose an NpgsqlReplicationAsyncEnumerable (I prefer that over Stream, because that's already taken); that could have a property pointing back to the connection, and even LSN properties which simply delegate back to it. This way, the user doesn't have to care about the intermediate enumerable, but they have the option of using it if they want to.

What do you think?

**Comment by Brar** [comment]
> Isn't cancellation handled via the cancellation token given to StartReplication? We don't support any sync I/O here, so I'm not sure we need a Cancel method...

No, we can absolutely only support cancellation via `CancellationToken` and completely discard the cancel method. It's perhaps a bit less obvious/discoverable but IMHO that doesn't outweigh the overhead of having two ways to cancel. I'm all for getting rid of `Cancel()`.

> Now, the way I'd expect 95% of users to consume this API, is as follows

I agree that `await foreach` is nice but if you look at the `StreamingLoop()` methods for variant 1 and 2 I'd say that the difference is really minor.

> in a scenario where the user wants to pass the enumerable around functions

This is actually close to 100% of the use cases.
The replication stream will probably always have to run in a separate thread since it may be cancelled by the server when it stops streaming (we go pretty far to avoid that with our automated feedback timer but the server may request feedback at any time and if this request isn't processed and the feedback sent accordingly, the server will cut the connection.

> NpgsqlReplicationAsyncEnumerable (I prefer that over Stream, because that's already taken)

You mean the term `NpgsqlReplicationStream` is a bad choice for an `IAsyncEnumerable<T>` because the name `Stream` is already taken?
I agree and initially shared this point of view, but then again there's [Async Streams](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/async-streams) and how would you convey [*streaming replication messages*](https://www.postgresql.org/docs/current/protocol-replication.html) if not via a replication stream?

> To address this, we could just have the best of both worlds: have StartReplication expose an NpgsqlReplicationAsyncEnumerable (I prefer that over Stream, because that's already taken); that could have a property pointing back to the connection, and even LSN properties which simply delegate back to it. This way, the user doesn't have to care about the intermediate enumerable, but they have the option of using it if they want to.

This feels like taking the implementation overhead without making the Interface slimmer or more concise.
I think that this way we're gathering more disadvantages than advantages.
I hate redundancy and if we have an API on the Enumerable I don't see why we should also have it on the connection or the other way around.

**Comment by roji** [comment]
> I agree that await foreach is nice but if you look at the StreamingLoop() methods for variant 1 and 2 I'd say that the difference is really minor.

At least in theory, there's also composition of LINQ operators (mainly Where, I think) which is only possible over enumerables.

>> in a scenario where the user wants to pass the enumerable around functions
>
> This is actually close to 100% of the use cases.
> The replication stream will probably always have to run in a separate thread since it may be cancelled by the server when it stops streaming (we go pretty far to avoid that with our automated feedback timer but the server may request feedback at any time and if this request isn't processed and the feedback sent accordingly, the server will cut the connection.

I'm probably missing some context here, but I don't see why server-initiated cancellation means that the *user code* doing the loop needs to go into a separate thread etc... From the user's perspective, say I'm writing a console program that listens to PG events and dumps them to some file. I'd expect to be able to do a simple `await foreach` loop in my Main, and if cancellation occurs I'd just expect a simple exception, no?

>> NpgsqlReplicationAsyncEnumerable (I prefer that over Stream, because that's already taken)
>
> You mean the term NpgsqlReplicationStream is a bad choice for an IAsyncEnumerable<T> because the name Stream is already taken?

Yeah, I'm only referring to naming here - NpgsqlReplicationStream wouldn't be a .NET Stream.

> I agree and initially shared this point of view, but then again there's Async Streams and how would you convey streaming replication messages if not via a replication stream?

I think Async Streams is a conceptual name for the C# feature, but AFAIK there aren't any .NET types using that name. The word iterator is also used in various places in the language specs ("iterator blocks") but not in any actual types etc.

>  I hate redundancy and if we have an API on the Enumerable I don't see why we should also have it on the connection or the other way around.

The point is to not force you to be aware of the enumerable in any way, in order to keep the API simple. My main goal here is to allow productive use of the API while directly doing `await foreach` over StartReplication. If you don't like the redundancy of having the same property on the enumerable and on the connection, we can just expose a simple reference from the enumerable to the connection - this would allow users to pass around the enumerable if they need to, and to get back to the connection from it.

**Comment by Brar** [comment]
> At least in theory, there's also composition of LINQ operators (mainly Where, I think) which is only possible over enumerables.

I didn't think about this.
Linq over streaming replication. That would be really...brave.
But why not. I'll note it for my tests.

> we can just expose a simple reference from the enumerable to the connection

I agree that this would be the sanest way but OTOH I probably wouldn't want to do an explicit IAsyncEnumerable implementation just because of a reference to the connection.

I think I'll see how far I get with the automatic IAsyncEnumerable.

Thanks for your opinion.

**Comment by roji** [comment]
>  Linq over streaming replication. That would be really...brave.

Yeah :) I can see myself using Where to functionally filter out only messages I want, not sure about any other operator. The peculiarity with replication is that the enumerable is infinite (I think?), so any terminating (non-lazy) operator is unusable.

On a totally unrelated note, that does raises the question of termination of the enumeration. I'm guessing that the only way to terminate is to cancel, in which case an OperationCancelledException is thrown (as is standard) - is that right? Or is there some scenario where we reach the end of the enumeration (e.g. can we ask to replicate from point A to B only, at which point the enumeration would simply complete)?

> I agree that this would be the sanest way but OTOH I probably wouldn't want to do an explicit IAsyncEnumerable implementation just because of a reference to the connection.

Up to you. I'm also not sure it's needed, but if passing around the enumerable (as opposed to the connection) really is something that will happen a lot, then this might make things slightly easier for people. You're the one best-placed to know.

**Comment by Brar** [comment]
> so any terminating (non-lazy) operator is unusable.

There's even more to it because the messages themselves contain a non-seekable stream that is still potentially wrapping the network stream, so you not only have to process the messages in order, you also can't access the data of a previous message once you've moved on to the next one.

> I'm guessing that the only way to terminate is to cancel, in which case an OperationCancelledException is thrown (as is standard) - is that right?

For the frontend program, yes.

> Or is there some scenario where we reach the end of the enumeration (e.g. can we ask to replicate from point A to B only, at which point the enumeration would simply complete)?

This can only happen if you start streaming right at the end of a timeline.
That's a corner case scenario that I still haven't tested yet, because of the complex conditions you need to set up to even prepare that state, but I already know how to induce timeline switches so in principle I know how to test it.

> if passing around the enumerable (as opposed to the connection) really is something that will happen a lot, then this might make things slightly easier for people

I completely agree, but changing an implicit IAsyncEnumerabe to an explicit implementation and adding properties to it is something we can do even after releasing since it wouldn't break the contract.

**Comment by roji** [comment]
>> so any terminating (non-lazy) operator is unusable.
>
> There's even more to it because the messages themselves contain a non-seekable stream that is still potentially wrapping the network stream, so you not only have to process the messages in order, you also can't access the data of a previous message once you've moved on to the next one.

IIRC, in the non-raw mode with that default plugin (I forgot what it's called), the user does get higher-level messages (DELETE, COMMIT), which can be accessed even after the enumerator has been advanced no? In this mode (which I expect everyone will use), some LINQ fanciness could be possible:

```c#
while (true) {
    var transactionMessages = replicationEnumerable.TakeWhile(m => !(m is CommitMessage)).ToListAsync();
    // Process a single transaction
}
```

Isn't that cool?

But otherwise, for the "raw mode" with XLogData, you're absolutely right. This isn't the first time I run into this issue with the "enumerable" abstraction - sometimes it represents a simple, full set of values in memory (so enumerating twice is OK, and all returns elements live and can be used at the same time), and sometimes it represents a stream-like thing where the moment you call Next on the enumerator, the previous value is no longer accessible. IMHO this is a shortcoming of the abstraction itself - this distinction should ideally have been represented via .NET types.

But beyond this fundamental problem with the abstraction, I don't think this specific replication scenario is very special - enumerables frequently represent data coming from the network and have the same gotchas as we have. Users just have to be aware of them.

:+1: on all the rest!
